### DISCUSSION TITLE
Length of Array in custom DataType

### DISCUSSION THREAD

**Discussion by AndreyKl4** [FreeOpcUa/opcua-asyncio#468] [Q&A]
Hi all,

i am a bit confused in what way i can set the length of an Array when i create a custom DataType with help DataTypeDeffinitions on a Server side.  When I use this line- I create a structure field that is an array but the length is not defined.

new_struct_field(field_name, ua_field[0], array=True)

When I connect to the structure, that includes mentioned above field, with help UaExpert I can select the ArraySize manually but i want to predefine  it when i create a custom DataType.

Thanks in Advance!

Andrey

**Comment by oroulet** [comment]
internally the array size is infinite in ua. But maybe there is an extra node property. Where do you find the size?

**Comment by AndreasHeine** [comment]
https://reference.opcfoundation.org/v104/Core/docs/Part3/5.6.2/

https://github.com/FreeOpcUa/opcua-asyncio/blob/12c0e7a05d756219ac56448edf7f78a030a3d049/asyncua/ua/uaprotocol_auto.py#L3278
```python

    field = ua.StructureField()
    field.Name = <name>
    field.IsOptional = False
    field.Description = ua.LocalizedText(text="TEST")
    field.DataType = ua.NodeId(<dtype>)
    field.ValueRank = ua.ValueRank.OneOrMoreDimensions
    field.ArrayDimensions = [<arraylength>]
```

**Reply by oroulet** [reply]
oh yeah that one. Looks like the interface is currently a bit too high level. It was an error to define array as bool. I probably copied it from somewhere else. We could have given the array valuerank and arraydimenstions directly...

**Reply by oroulet** [reply]
maybe we can just add them as-is

**Reply by AndreyKl4** [reply]
@oroulet  @AndreasHeine thank for your replies. I did what @AndreasHeine mentioned above. 

> field = ua.StructureField()
> field.Name = "tags"
> field.IsOptional = False
> field.Description = ua.LocalizedText(text="TEST")
> field.DataType = ua.NodeId(my_data_type)
> field.ValueRank = ua.ValueRank.OneOrMoreDimensions
> field.ArrayDimensions = [4]

After that i checked the behavior in UaExpert and it is differ from behavior in Spec https://reference.opcfoundation.org/v104/Core/docs/Part3/5.6.2/.  In screenshot u will see that array is UInt32 Array[1] but it suppose to  be UInt32 Array[4] as far as I correct understood the Spec. Also i tried to use Siome as client to monitor ArrayDimensions of DataType in Siome the Atribute "ArrayDimensions" is empty. Both screens i have already attached.
 

Thanks in Advance!

<img width="179" alt="ua_expert_screenshot" src="https://user-images.githubusercontent.com/73280686/110467563-78539680-80d7-11eb-8169-5847c671f242.PNG">

<img width="491" alt="siome_screenshot" src="https://user-images.githubusercontent.com/73280686/110468069-121b4380-80d8-11eb-9535-898ce7a18274.PNG">
